<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform goseq Enrichment tests across a GeneSetDb. — goseq • sparrow</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Perform goseq Enrichment tests across a GeneSetDb. — goseq"><meta property="og:description" content="Note that we do not import things from goseq directly, and only load
it if this function is fired. I can't figure out a way to selectively
import functions from the goseq package without it having to load its
dependencies, which take a long time -- and I don't want loading sparrow
to take a long time. So, the goseq package has moved to Suggests and then
is loaded within this function when necessary."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sparrow</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/sparrow.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/WIP/adding-new-methods-to-sparrow.html">Adding new methods to sparrow</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lianos/sparrow/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform goseq Enrichment tests across a GeneSetDb.</h1>
    <small class="dont-index">Source: <a href="https://github.com/lianos/sparrow/blob/HEAD/R/do.goseq.R" class="external-link"><code>R/do.goseq.R</code></a></small>
    <div class="hidden name"><code>goseq.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Note that we do not import things from goseq directly, and only load
it if this function is fired. I can't figure out a way to selectively
import functions from the goseq package without it having to load its
dependencies, which take a long time -- and I don't want loading sparrow
to take a long time. So, the goseq package has moved to Suggests and then
is loaded within this function when necessary.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">goseq</span><span class="op">(</span></span>
<span>  <span class="va">gsd</span>,</span>
<span>  <span class="va">selected</span>,</span>
<span>  <span class="va">universe</span>,</span>
<span>  <span class="va">feature.bias</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Wallenius"</span>, <span class="st">"Sampling"</span>, <span class="st">"Hypergeometric"</span><span class="op">)</span>,</span>
<span>  repcnt <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  use_genes_without_cat <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot.fit <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  do.conform <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  as.dt <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  .pipelined <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gsd</dt>
<dd><p>The <code>GeneSetDb</code> object to run tests against</p></dd>


<dt>selected</dt>
<dd><p>The ids of the selected features</p></dd>


<dt>universe</dt>
<dd><p>The ids of the universe</p></dd>


<dt>feature.bias</dt>
<dd><p>a named vector as long as <code>nrow(x)</code> that has the
"bias" information for the features/genes tested (ie. vector of gene
lengths). <code>names(feature.bias)</code> should equal <code>rownames(x)</code>.
If this is not provided, all feature lengths are set to 1 (no bias).
The goseq package provides a <code><a href="https://rdrr.io/pkg/goseq/man/getlength.html" class="external-link">getlength</a></code> function which
facilitates getting default values for these if you do not have the
correct values used in your analysis.</p></dd>


<dt>method</dt>
<dd><p>The method to use to calculate the unbiased category
enrichment scores</p></dd>


<dt>repcnt</dt>
<dd><p>Number of random samples to be calculated when random sampling
is used. Ignored unless <code>method="Sampling"</code>.</p></dd>


<dt>use_genes_without_cat</dt>
<dd><p>A boolean to indicate whether genes without a
categorie should still be used. For example, a large number of gene may
have no GO term annotated. If this option is set to FALSE, those genes
will be ignored in the calculation of p-values (default behaviour). If
this option is set to TRUE, then these genes will count towards the total
number of genes outside the category being tested.</p></dd>


<dt>plot.fit</dt>
<dd><p>parameter to pass to <code><a href="https://rdrr.io/pkg/goseq/man/nullp.html" class="external-link">goseq::nullp()</a></code>.</p></dd>


<dt>do.conform</dt>
<dd><p>By default <code>TRUE</code>: does some gymnastics to conform
the <code>gsd</code> to the <code>universe</code> vector. This should neber be set
to <code>FALSE</code>, but this parameter is here so that when this function
is called from the <code><a href="seas.html">seas()</a></code> codepath, we do not have to
reconform the <code>GeneSetDb</code> object, because it has already been done.</p></dd>


<dt>as.dt</dt>
<dd><p>If <code>FALSE</code> (default), the data.frame like thing that
this funciton returns will be set to a data.frame. Set this to <code>TRUE</code>
to keep this object as a <code>data.table</code></p></dd>


<dt>.pipelined</dt>
<dd><p>If this is being called external to a seas pipeline, then
some additional cleanup of columns name output will be done when
<code>FALSE</code> (default). Otherwise the column renaming and post processing is
left to the do.goseq caller.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A <code>data.table</code> of results, similar to goseq output. The output
from <code><a href="https://rdrr.io/pkg/goseq/man/nullp.html" class="external-link">nullp</a></code> is added to the outgoing data.table as
an attribue named <code>"pwf"</code>.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Young, M. D., Wakefield, M. J., Smyth, G. K., Oshlack, A. (2010).
Gene ontology analysis for RNA-seq: accounting for selection bias.
<em>Genome Biology</em> 11, R14. http://genomebiology.com/2010/11/2/R14</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">vm</span> <span class="op">&lt;-</span> <span class="fu"><a href="examples.html">exampleExpressionSet</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="conform.html">conform</a></span><span class="op">(</span><span class="fu"><a href="examples.html">exampleGeneSetDb</a></span><span class="op">(</span><span class="op">)</span>, <span class="va">vm</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Identify DGE genes</span></span></span>
<span class="r-in"><span><span class="va">mg</span> <span class="op">&lt;-</span> <span class="fu"><a href="seas.html">seas</a></span><span class="op">(</span><span class="va">vm</span>, <span class="va">gdb</span>, design <span class="op">=</span> <span class="va">vm</span><span class="op">$</span><span class="va">design</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">lfc</span> <span class="op">&lt;-</span> <span class="fu"><a href="logFC.html">logFC</a></span><span class="op">(</span><span class="va">mg</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># wire up params</span></span></span>
<span class="r-in"><span><span class="va">selected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">lfc</span>, <span class="va">significant</span><span class="op">)</span><span class="op">$</span><span class="va">feature_id</span></span></span>
<span class="r-in"><span><span class="va">universe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">vm</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mylens</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">vm</span><span class="op">$</span><span class="va">genes</span><span class="op">$</span><span class="va">size</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">vm</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">degenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">universe</span><span class="op">)</span><span class="op">)</span>, <span class="va">universe</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">degenes</span><span class="op">[</span><span class="va">selected</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1L</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gostats</span> <span class="op">&lt;-</span> <span class="fu">sparrow</span><span class="fu">::</span><span class="fu">goseq</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">gdb</span>, <span class="va">selected</span>, <span class="va">universe</span>, <span class="va">mylens</span>,</span></span>
<span class="r-in"><span>  method <span class="op">=</span> <span class="st">"Wallenius"</span>, use_genes_without_cat <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>initial point very close to some inequality constraints</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using manually entered categories.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating the p-values...</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Steve Lianoglou, Denali Therapeutics, Genentech.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

