<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Creates a "geneset smart" ComplexHeatmap::Heatmap — mgheatmap • sparrow</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Creates a " geneset smart complexheatmap::heatmap mgheatmap><meta property="og:description" content="Before we get started, note that you probably want to use mgheatmap2().
This function encapsulates many common &quot;moves&quot; you'll make when trying to
make a heatmap, especially if you are trying to show geneset activity across
a panel of samples.
NOTE: this function will almost certainly reorder the rows of the
input matrix. If you are concatentating Heatmap objects together horizontally
(ie. you if you want to use a rowAnnotation along side the returned heatmap),
you must reorder the rows of the annotation data.frame, ie.
ranno.df &amp;lt;- ranno.df[rownames(out@matrix),]"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sparrow</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.11.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/sparrow.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/WIP/adding-new-methods-to-sparrow.html">Adding new methods to sparrow</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lianos/sparrow/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Creates a "geneset smart" ComplexHeatmap::Heatmap</h1>
    <small class="dont-index">Source: <a href="https://github.com/lianos/sparrow/blob/devel/R/plots-mgheatmap.R" class="external-link"><code>R/plots-mgheatmap.R</code></a></small>
    <div class="hidden name"><code>mgheatmap.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Before we get started, note that you probably want to use <code><a href="mgheatmap2.html">mgheatmap2()</a></code>.</p>
<p>This function encapsulates many common "moves" you'll make when trying to
make a heatmap, especially if you are trying to show geneset activity across
a panel of samples.</p>
<p><strong>NOTE</strong>: this function will <strong>almost certainly</strong> reorder the rows of the
input matrix. If you are concatentating Heatmap objects together horizontally
(ie. you if you want to use a rowAnnotation along side the returned heatmap),
you must reorder the rows of the annotation data.frame, ie.
<code>ranno.df &lt;- ranno.df[rownames(out@matrix),]</code></p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mgheatmap</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  gdb <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  aggregate.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"none"</span>, <span class="st">"ewm"</span>, <span class="st">"ewz"</span>, <span class="st">"zscore"</span><span class="op">)</span>,</span>
<span>  split <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scores <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gs.order <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rm.collection.prefix <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rm.dups <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  recenter <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rescale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rename.rows <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  zero_center_colramp <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  zlim <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transpose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>the data matrix</p></dd>


<dt id="arg-gdb">gdb<a class="anchor" aria-label="anchor" href="#arg-gdb"></a></dt>
<dd><p><code>GeneSetDb</code> object that holds the genesets to plot. Defaults to
<code>NULL</code>, which will plot all rows in <code>x</code>.</p></dd>


<dt id="arg-col">col<a class="anchor" aria-label="anchor" href="#arg-col"></a></dt>
<dd><p>a colorRamp(2) function</p></dd>


<dt id="arg-aggregate-by">aggregate.by<a class="anchor" aria-label="anchor" href="#arg-aggregate-by"></a></dt>
<dd><p>the method used to generate single-sample geneset
scores. Default is <code>none</code> which plots heatmap at the gene level</p></dd>


<dt id="arg-split">split<a class="anchor" aria-label="anchor" href="#arg-split"></a></dt>
<dd><p>introduce row-segmentation based on genesets or collections?
Defaults is <code>TRUE</code> which will create split heatmaps based on
collection if <code>aggregate.by != 'none'</code>, or based on gene sets
if <code>aggregate.by == "none"</code>.</p></dd>


<dt id="arg-scores">scores<a class="anchor" aria-label="anchor" href="#arg-scores"></a></dt>
<dd><p>If <code>aggregate.by != "none"</code> you can pass in a precomupted
<code><a href="scoreSingleSamples.html">scoreSingleSamples()</a></code> result, otherwise one will be
computed internally. Note that if this is a <code>data.frame</code> of
pre-computed scores, the <code>gdb</code> is largely irrelevant (but still
required).</p></dd>


<dt id="arg-gs-order">gs.order<a class="anchor" aria-label="anchor" href="#arg-gs-order"></a></dt>
<dd><p>This is experimental, and is here to help order the order
of the genesets (or genesets collection) in a different way than the
default. By default, <code>gs.order = NULL</code> and genesets are enumerated in
alphabetical in the heatmap. You can pass in a character vector that will
dictate the order of the genesets displayed in the heatmap. Currently this
only matches against the <code>"name"</code> value of the geneset and probably only
works when <code>split = TRUE</code>. We will support <code>colleciton,name</code> tuples soon.
This can be a superset of the names found in <code>gdb</code>. As of ComplexHeatmap
v2 (maybe earlier versions), this doesn't really work when
<code>cluster_rows = TRUE</code>.</p></dd>


<dt id="arg-name">name<a class="anchor" aria-label="anchor" href="#arg-name"></a></dt>
<dd><p>passed down to <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code></p></dd>


<dt id="arg-rm-collection-prefix">rm.collection.prefix<a class="anchor" aria-label="anchor" href="#arg-rm-collection-prefix"></a></dt>
<dd><p>When <code>TRUE</code> (default), removes the collection
name from the genesets annotated on the heatmap.</p></dd>


<dt id="arg-rm-dups">rm.dups<a class="anchor" aria-label="anchor" href="#arg-rm-dups"></a></dt>
<dd><p>if <code>aggregate.by == 'none'</code>, do we remove genes that
appear in more than one geneset? Defaults to <code>FALSE</code></p></dd>


<dt id="arg-recenter">recenter<a class="anchor" aria-label="anchor" href="#arg-recenter"></a></dt>
<dd><p>do you want to mean center the rows of the heatmap matrix
prior to calling <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>?</p></dd>


<dt id="arg-rescale">rescale<a class="anchor" aria-label="anchor" href="#arg-rescale"></a></dt>
<dd><p>do you want to standardize the row variance to one on the
values of the heatmap matrix prior to calling
<code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>?</p></dd>


<dt id="arg-center-scale">center, scale<a class="anchor" aria-label="anchor" href="#arg-center-scale"></a></dt>
<dd><p>boolean parameters passed down into the the single
sample gene set scoring methods defined by <code>aggregate.by</code></p></dd>


<dt id="arg-rename-rows">rename.rows<a class="anchor" aria-label="anchor" href="#arg-rename-rows"></a></dt>
<dd><p>defaults to <code>NULL</code>, which induces no action. Specifying
a paramter here assumes you want to rename the rows of the heatmap.
Please refer to the "Renaming Rows" section for details.</p></dd>


<dt id="arg-zero-center-colramp">zero_center_colramp<a class="anchor" aria-label="anchor" href="#arg-zero-center-colramp"></a></dt>
<dd><p>Used to specify the type of color ramp to generate
when <code>col</code> is <code>NULL</code>. By default (<code>NULL</code>) we try to guess if we should
generate a 0-centered (blue, white, red) color ramp, or an absolute
(viridis style) one. The guessing functionality isn't that great, so
it doesn't hurt to explicitly set this to <code>TRUE</code> or <code>FALSE</code>.</p></dd>


<dt id="arg-zlim">zlim<a class="anchor" aria-label="anchor" href="#arg-zlim"></a></dt>
<dd><p>Used to control the color saturation of the heatmap when the
<code>col</code> parameter is not provided. If <code>NULL</code>, (default), extreme values
(outside the <code>c(0.025, 0.975)</code> quantiles) are axed and the colorRamp is
based on the remaining value range. If <code>FALSE</code>, the range of the colorRamp
is defined by the min/max values. Otherwise a length(2) numeric can be
supplied. If the values are between <code>[0,1]</code>, then we assume this is a
quantile range to be calculated. Otherwise the number are assumed to
mark the top and bottom of the color scale range you want to use.</p></dd>


<dt id="arg-transpose">transpose<a class="anchor" aria-label="anchor" href="#arg-transpose"></a></dt>
<dd><p>Flip display so that rows are columns. Default is <code>FALSE</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>parameters to send down to <code><a href="scoreSingleSamples.html">scoreSingleSamples()</a></code>,
<code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>, <code><a href="renameRows.html">renameRows()</a></code> internal <code>as_matrix()</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code>Heatmap</code> object.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>More info here.</p>
    </div>
    <div id="renaming-heatmap-rows">
    <h2>Renaming Heatmap Rows</h2>


<p>This function leverages <code><a href="renameRows.html">renameRows()</a></code> so that you can better customize the
output of your heatmaps by tweaking its rownames.</p>
<p>If you are plotting a <strong>gene-level</strong> heatmap (ie. <code>aggregate.by == "none"``) and the </code>rownames()<code>are gene identifiers, but you want the rownames of the heatmap to be gene symbols. You can perform this renaming using the</code>rename.rows` parameter.</p><ul><li><p>If <code>rename.rows</code> is <code>NULL</code>, then nothing is done.</p></li>
<li><p>If <code>rename.rows</code> is a <code>string</code>, then we assume that <code>x</code> has an associated
metadata <code>data.frame</code> over its rows and that <code>rename.rows</code> names one of
its columns, ie. <code>DGEList$genes[[rename.rows]]</code> or
<code>fData(ExpressionSet)[[rename.rows]]</code>. The values in that column will
be swapped out for <code>x</code>'s rownames</p></li>
<li><p>If <code>rename.rows</code> is a two-column data.frame, the first column is assumed
to be <code>rownames(x)</code> and the second is what you want to rename it to.</p></li>
<li><p>When there are duplicates in the renamed rownames, the <code>rename.duplicates</code>
<code>...</code> parameter dictates the behavior. This will happen, for instance, if
you are trying to rename the rows of an affy matrix to gene symbols, where
we have multiple probe ids for one gene. When <code>rename.duplicates</code> is set to
<code>"original"</code>, one of the rows will get the new name, and the remaning
duplicate rows will keep the rownames they came in with. When set to
<code>"make.unique"</code>, the new names will contain <code>*.1</code>, <code>*.2</code>, etc. suffixes,
as you would get from using <code><a href="https://rdrr.io/r/base/make.unique.html" class="external-link">base::make.unique()</a></code>.</p></li>
</ul><p>Maybe you are aggregating the expression scores into geneset scores, and
you don't want the rownames of the heatmap to be <code>collection;;name</code> (or just
<code>name</code> when <code>rm.collection.prefx = TRUE</code>), you can pass in a two column
<code>data.frame</code>, where the first column is <code>collection;name</code> and the second
is the name you want to rename that to. There is an example of this in
the "Examples" section here.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="mgheatmap2.html">mgheatmap2()</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: grid</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ========================================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ComplexHeatmap version 2.21.1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Github page: https://github.com/jokergoo/ComplexHeatmap</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Documentation: http://jokergoo.github.io/ComplexHeatmap-reference</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> If you use it in published research, please cite either one:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> - Gu, Z. Complex Heatmap Visualization. iMeta 2022.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> - Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     genomic data. Bioinformatics 2016.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The new InteractiveComplexHeatmap package can directly export static </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> complex heatmaps into an interactive Shiny app with zero effort. Have a try!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> This message can be suppressed by:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   suppressPackageStartupMessages(library(ComplexHeatmap))</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ========================================</span>
<span class="r-in"><span><span class="va">vm</span> <span class="op">&lt;-</span> <span class="fu"><a href="examples.html">exampleExpressionSet</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="examples.html">exampleGeneSetDb</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">col.anno</span> <span class="op">&lt;-</span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html" class="external-link">HeatmapAnnotation</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  df <span class="op">=</span> <span class="va">vm</span><span class="op">$</span><span class="va">targets</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cancer_Status"</span>, <span class="st">"PAM50subtype"</span><span class="op">)</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    Cancer_Status <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>normal <span class="op">=</span> <span class="st">"grey"</span>, tumor <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    PAM50subtype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Basal <span class="op">=</span> <span class="st">"purple"</span>, Her2 <span class="op">=</span> <span class="st">"green"</span>, LumA <span class="op">=</span> <span class="st">"orange"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mgh</span> <span class="op">&lt;-</span> <span class="fu">mgheatmap</span><span class="op">(</span><span class="va">vm</span>, <span class="va">gdb</span>, aggregate.by <span class="op">=</span> <span class="st">"ewm"</span>, split<span class="op">=</span><span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                 top_annotation <span class="op">=</span> <span class="va">col.anno</span>, show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>                 column_title <span class="op">=</span> <span class="st">"Gene Set Activity in BRCA subset"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Maybe you want the rownames of the matrix to use spaces instead of "_"</span></span></span>
<span class="r-in"><span><span class="va">rr</span> <span class="op">&lt;-</span> <span class="fu"><a href="geneSets.html">geneSets</a></span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span><span class="op">[</span>, <span class="st">"name"</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">rr</span><span class="op">$</span><span class="va">newname</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_"</span>, <span class="st">" "</span>, <span class="va">rr</span><span class="op">$</span><span class="va">name</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mg2</span> <span class="op">&lt;-</span> <span class="fu">mgheatmap</span><span class="op">(</span><span class="va">vm</span>, <span class="va">gdb</span>, aggregate.by<span class="op">=</span><span class="st">'ewm'</span>, split<span class="op">=</span><span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                 top_annotation <span class="op">=</span> <span class="va">col.anno</span>, show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>                 column_title <span class="op">=</span> <span class="st">"Gene Set Activity in BRCA subset"</span>,</span></span>
<span class="r-in"><span>                 rename.rows <span class="op">=</span> <span class="va">rr</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Steve Lianoglou, Aratus Informatics, LLC, Denali Therapeutics, Genentech.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

