<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Performs an overrepresentation analysis, (optionally) accounting for bias. — ora • sparrow</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Performs an overrepresentation analysis, (optionally) accounting for bias. — ora"><meta property="og:description" content="This function wraps limma::kegga() to perform biased overrepresntation
analysis over gene set collection stored in a GeneSetDb (gsd) object. Its
easiest to use this function when the biases and selection criteria are
stored as columns of the input data.frame dat."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sparrow</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/sparrow.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/WIP/adding-new-methods-to-sparrow.html">Adding new methods to sparrow</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lianos/sparrow/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Performs an overrepresentation analysis, (optionally) accounting for bias.</h1>
    <small class="dont-index">Source: <a href="https://github.com/lianos/sparrow/blob/HEAD/R/do.ora.R" class="external-link"><code>R/do.ora.R</code></a></small>
    <div class="hidden name"><code>ora.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function wraps <code><a href="https://rdrr.io/pkg/limma/man/goana.html" class="external-link">limma::kegga()</a></code> to perform biased overrepresntation
analysis over gene set collection stored in a GeneSetDb (<code>gsd</code>) object. Its
easiest to use this function when the biases and selection criteria are
stored as columns of the input data.frame <code>dat</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ora</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">gsd</span>,</span>
<span>  selected <span class="op">=</span> <span class="st">"significant"</span>,</span>
<span>  groups <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  feature.bias <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  universe <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  restrict.universe <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plot.bias <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  as.dt <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">plot_ora_bias</span><span class="op">(</span><span class="va">x</span>, <span class="va">selected</span>, <span class="va">feature.bias</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>A data.frame with feature-level statistics. Minimally, this should
have a <code>"feature_id"</code> (character) column, but read on ...</p></dd>


<dt>gsd</dt>
<dd><p>The GeneSetDb</p></dd>


<dt>selected</dt>
<dd><p>Either the name of a logical column in <code>dat</code> used to subset
out the features to run the enrichement over, or a character vector of
<code>"feature_id"</code>s that are selected from <code>dat[["feature_id"]]</code>.</p></dd>


<dt>groups</dt>
<dd><p>Encodes groups of features that we can use to test selected
features individual, as well as "all" together. This can be specified by:
(1) specifying a name of a column in <code>dat</code> to split the enriched features
into subgroups. (2) A named list of features to intersect with <code>selected</code>.
By default this is <code>NULL</code>, so we only run enrichment over
all elements in <code>selected</code>. See examples for details.</p></dd>


<dt>feature.bias</dt>
<dd><p>If <code>NULL</code> (default), no bias is used in enrichment
analysis. Otherwise, can be the name of a column in <code>dat</code> to extract
a numeric bias vector (gene length, GC content, average expression, etc.)
or a named (using featureIds) numeric vector of the same. The BiasedUrn
CRAN package is required when this is not NULL.</p></dd>


<dt>universe</dt>
<dd><p>Defaults to all elements in <code>dat[["feature_id"]]</code>.</p></dd>


<dt>restrict.universe</dt>
<dd><p>See same parameter in <code><a href="https://rdrr.io/pkg/limma/man/goana.html" class="external-link">limma::kegga()</a></code></p></dd>


<dt>plot.bias</dt>
<dd><p>See <code>plot</code> parameter in <code><a href="https://rdrr.io/pkg/limma/man/goana.html" class="external-link">limma::kegga()</a></code>. You can generate
this plot without running <code>ora</code> using the <code>plot_ora_bias()</code>,
like so:
<code>plot_ora_bias(dat, selected = selected, groups = groups, feature.bias = feature.bias)</code></p></dd>


<dt>...</dt>
<dd><p>parameters passed to <code><a href="conform.html">conform()</a></code></p></dd>


<dt>as.dt</dt>
<dd><p>If <code>FALSE</code> (default), the data.frame like thing that
this funciton returns will be set to a data.frame. Set this to <code>TRUE</code>
to keep this object as a <code>data.table</code></p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data.frame of pathway enrichment. The last N colums are enrichment
statistics per pathway, grouped by the <code>groups</code> parameter. <code>P.all</code> are the
stats for all selected features, and the remaingin <code>P.*</code> columns are for
the features specifed by <code>groups</code>.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>In principle, this test does what <code>goseq</code> does, however I found that
sometimes calling goseq would throw errors within <code><a href="https://rdrr.io/pkg/goseq/man/nullp.html" class="external-link">goseq::nullp()</a></code> when
calling <code>makesplines</code>. I stumbled onto this implementation when googling
for these errors and landing here:
https://support.bioconductor.org/p/65789/#65914</p>
<p>The meat and potatoes of this function's code was extracted from
<code><a href="https://rdrr.io/pkg/limma/man/goana.html" class="external-link">limma::kegga()</a></code>, written by Gordon Smyth and Yifang Hu.</p>
<p>Note that the BiasedUrn CRAN package needs to be installed to support biased
enrichment testing</p>
    </div>
    <div id="functions">
    <h2>Functions</h2>
    
<ul><li><p><code>plot_ora_bias()</code>: plots the bias of coviarate to DE / selected status. Code
taken from <code><a href="https://rdrr.io/pkg/limma/man/goana.html" class="external-link">limma::kegga()</a></code></p></li>
</ul></div>
    <div id="references">
    <h2>References</h2>
    <p>Young, M. D., Wakefield, M. J., Smyth, G. K., Oshlack, A. (2010).
Gene ontology analysis for RNA-seq: accounting for selection bias.
<em>Genome Biology</em> 11, R14. http://genomebiology.com/2010/11/2/R14</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">dgestats</span> <span class="op">&lt;-</span> <span class="fu"><a href="examples.html">exampleDgeResult</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="randomGeneSetDb.html">randomGeneSetDb</a></span><span class="op">(</span><span class="va">dgestats</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run enrichmnent without accounting for any bias</span></span></span>
<span class="r-in"><span><span class="va">nobias</span> <span class="op">&lt;-</span> <span class="fu">ora</span><span class="op">(</span><span class="va">dgestats</span>, <span class="va">gdb</span>, selected <span class="op">=</span> <span class="st">"selected"</span>, groups <span class="op">=</span> <span class="st">"direction"</span>,</span></span>
<span class="r-in"><span>              feature.bias <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run enrichment and account for gene length</span></span></span>
<span class="r-in"><span><span class="va">lbias</span> <span class="op">&lt;-</span> <span class="fu">ora</span><span class="op">(</span><span class="va">dgestats</span>, <span class="va">gdb</span>, selected <span class="op">=</span> <span class="st">"selected"</span>,</span></span>
<span class="r-in"><span>             feature.bias <span class="op">=</span> <span class="st">"effective_length"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># plot length bias with DGE status</span></span></span>
<span class="r-in"><span><span class="fu">plot_ora_bias</span><span class="op">(</span><span class="va">dgestats</span>, <span class="st">"selected"</span>, <span class="st">"effective_length"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="ora-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># induce length bias and see what is the what ...............................</span></span></span>
<span class="r-in"><span><span class="va">biased</span> <span class="op">&lt;-</span> <span class="va">dgestats</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">dgestats</span><span class="op">$</span><span class="va">pval</span><span class="op">)</span>,<span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">biased</span><span class="op">$</span><span class="va">effective_length</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">biased</span><span class="op">$</span><span class="va">effective_length</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">plot_ora_bias</span><span class="op">(</span><span class="va">biased</span>, <span class="st">"selected"</span>, <span class="st">"effective_length"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="ora-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="va">etest</span> <span class="op">&lt;-</span> <span class="fu">ora</span><span class="op">(</span><span class="va">biased</span>, <span class="va">gdb</span>, selected <span class="op">=</span> <span class="st">"selected"</span>,</span></span>
<span class="r-in"><span>             groups <span class="op">=</span> <span class="st">"direction"</span>,</span></span>
<span class="r-in"><span>             feature.bias <span class="op">=</span> <span class="st">"effective_length"</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Steve Lianoglou, Aratus Informatics, LLC, Denali Therapeutics, Genentech.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

