<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Performing gene set enrichment analyses with sparrow • sparrow</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Performing gene set enrichment analyses with sparrow">
<meta property="og:description" content="sparrow">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sparrow</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.11.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/sparrow.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/WIP/adding-new-methods-to-sparrow.html">Adding new methods to sparrow</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lianos/sparrow/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Performing gene set enrichment analyses with
sparrow</h1>
                        <h4 data-toc-skip class="author">Steve
Lianoglou</h4>
            
            <h4 data-toc-skip class="date">26 May 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/lianos/sparrow/blob/HEAD/vignettes/sparrow.Rmd" class="external-link"><code>vignettes/sparrow.Rmd</code></a></small>
      <div class="hidden name"><code>sparrow.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>The sparrow package facilitates the use of gene sets in the
      analysis of high throughput genomics data. It provides simple
      execution and comparison of several GSEA approaches through a
      unified interface within the user’s workspace or interactively via
      a shiny application provided by the sparrow.shiny package. This
      package also provides an easy wrapper to single sample gene set
      scoring and geneset-centric heatmaps for visualization. sparrow
      package version: 1.11.0</p>
    </div>
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>The <a href="https://github.com/lianos/sparrow" class="external-link">sparrow</a> package was built to facilitate the use of
gene sets in the analysis of high throughput genomics data (primarily
RNA-seq). It does so by providing these top-line functionalities:</p>
<ul>
<li>The <code>seas</code> function is a wrapper that orchestrates the
execution of any number of user-specified gene set enrichment analyses
(GSEA) over a particular experimental contrast of interest. This will
create a <code>SparrowResult</code> object which stores the results of
each GSEA method internally, allowing for easy query and retrieval.</li>
<li>A sister <a href="https://github.com/lianos/sparrow.shiny" class="external-link"><code>{sparrow.shiny}</code></a>
package provides an <code>explore</code> function, which is invoked on
<code>SparrowResult</code> objects returned from a call to
<code>seas</code>. The shiny application facilitates interactive
exploration of these GSEA results. This application can also be deployed
to a shiny server and can be initialized by uploading a serialized
<code>SparrowResult</code> <code>*.rds</code> file.</li>
<li>An “over representation analysis” method <code><a href="../reference/ora.html">ora()</a></code> which
wraps the biased enrichment functionality found within
<code><a href="https://rdrr.io/pkg/limma/man/goana.html" class="external-link">limma::kegga</a></code> and generalizes it to work against data.frame
inputs with arbitrary genesets.</li>
<li>The <code>scoreSingleSamples</code> function is a wrapper that
enables the user to generate single sample gene set scores using a
variety of different single sample gene set scoring methods.</li>
<li>Convenience gene set collection retrieval functions that return
<code>BiocSet</code>s from widely used databases, like
<code><a href="../reference/getMSigCollection.html">getMSigCollection()</a></code> (<a href="http://software.broadinstitute.org/gsea/msigdb/" class="external-link">MSigDB</a>),
<code><a href="../reference/getKeggCollection.html">getKeggCollection()</a></code> (<a href="https://www.genome.jp/kegg/" class="external-link">KEGG</a>),
<code><a href="../reference/getPantherCollection.html">getPantherCollection()</a></code> (<a href="pantherdb">PANTHER
database</a>), and <code><a href="../reference/getReactomeCollection.html">getReactomeCollection()</a></code> (<a href="https://reactome.org/" class="external-link">Reactome</a>) with support for different
organisms and identifier types (partially).</li>
</ul>
<p>The initial GSEA methods that sparrow wrapped were the ones provided
by limma and edgeR. As such, many analyses using sparrow expect you to
re-use the same data objects used for differential expression analysis,
namely:</p>
<ul>
<li>Expression data (an <code>EList</code>, <code>DGEList</code>, or
expression matrix)</li>
<li>A design matrix</li>
<li>A contrast vector/matrix (if your design and comparison require
it)</li>
</ul>
<p>Other methods only require the user to provide a ranked vector of
statistics that represent some differential expression statistic per
gene, and the GSEA is performed by analyzing the ranks of genes within
this vector.</p>
<p>The user can invoke one <code><a href="../reference/seas.html">seas()</a></code> call that can orchestrate
multiple analyses of any type.</p>
<p>Currently supported gene set enrichment methods include:</p>
<pre><code><span><span class="co">##            method test_type package</span></span>
<span><span class="co">## 1          camera preranked   limma</span></span>
<span><span class="co">## 2        cameraPR preranked   limma</span></span>
<span><span class="co">## 3           fgsea preranked   fgsea</span></span>
<span><span class="co">## 4             ora       ora     ora</span></span>
<span><span class="co">## 5             fry preranked   limma</span></span>
<span><span class="co">## 6           roast preranked   limma</span></span>
<span><span class="co">## 7           romer preranked   limma</span></span>
<span><span class="co">## 8           goseq       ora   goseq</span></span>
<span><span class="co">## 9     geneSetTest preranked   limma</span></span>
<span><span class="co">## 10          logFC preranked   limma</span></span>
<span><span class="co">## 11 svdGeneSetTest      meta sparrow</span></span></code></pre>
<p>When using these methods in analyses that lead to publication, please
cite the original papers that developed these methods and cite sparrow
when its functionality assisted in your interpretation and analysis.</p>
<p>The sparrow package provides a small example expression dataset
extracted from the TCGA BRCA dataset, which is available via the
<code>exampleExpressionSet</code> function. In this vignette we will
explore differential expression and gene set enrichment analysis by
examining differences between basal and her2 PAM50 subtypes.</p>
</div>
<div class="section level2">
<h2 id="standard-workflow">Standard Workflow<a class="anchor" aria-label="anchor" href="#standard-workflow"></a>
</h2>
<p>Let’s begin by setting up our work environment for exploratory
analysis using the sparrow package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lianos/sparrow" class="external-link">sparrow</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize" class="external-link">circlize</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="note">
<p>Internally, sparrow leverages the <a href="https://CRAN.R-project.org/package=data.table" class="external-link">data.table</a>
package for fast indexing and manipulation over data.frames. All
functions that return data.frame looking objects back have converted it
from an data.table prior to return. All such functions take an
<code>as.dt</code> argument, which is set to <code>FALSE</code> by
default that controls this behavior. If you want <a href="https://github.com/lianos/sparrow" class="external-link">sparrow</a>
to return a data.table back to you from some function, try adding an
<code>as.dt = TRUE</code> argument to the end of the function call.</p>
</div>
<div class="section level3">
<h3 id="data-setup">Data Setup<a class="anchor" aria-label="anchor" href="#data-setup"></a>
</h3>
<p>sparrow is most straightforward to use when our data objects and
analysis are performed with either the edgeR or voom/limma pipelines and
when we use standard gene identifiers (like esnemble) as
<code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames()</a></code> to these objects.</p>
<p>The <code>exampleExpressionSet</code> function gives us just such an
object. We call it below in a manner that gives us an object that allows
us to explore expression differences between different subtypes of
breast cancer.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/examples.html">exampleExpressionSet</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"tumor-subtype"</span>, do.voom <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Below you’ll find the <code>$targets</code> data.frame of the voomed
<code>EList</code></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vm</span><span class="op">$</span><span class="va">targets</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Patient_ID</span>, <span class="va">Cancer_Status</span>, <span class="va">PAM50subtype</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                Patient_ID Cancer_Status PAM50subtype</span></span>
<span><span class="co">## TCGA-A2-A0CM-01A-31R-A034-07 TCGA-A2-A0CM         tumor        Basal</span></span>
<span><span class="co">## TCGA-BH-A0RX-01A-21R-A084-07 TCGA-BH-A0RX         tumor        Basal</span></span>
<span><span class="co">## TCGA-BH-A18Q-01A-12R-A12D-07 TCGA-BH-A18Q         tumor        Basal</span></span>
<span><span class="co">## TCGA-B6-A0RU-01A-11R-A084-07 TCGA-B6-A0RU         tumor        Basal</span></span>
<span><span class="co">## TCGA-BH-A18P-01A-11R-A12D-07 TCGA-BH-A18P         tumor         Her2</span></span>
<span><span class="co">## TCGA-C8-A275-01A-21R-A16F-07 TCGA-C8-A275         tumor         Her2</span></span>
<span><span class="co">## TCGA-C8-A12Z-01A-11R-A115-07 TCGA-C8-A12Z         tumor         Her2</span></span>
<span><span class="co">## TCGA-A2-A0T1-01A-21R-A084-07 TCGA-A2-A0T1         tumor         Her2</span></span>
<span><span class="co">## TCGA-AC-A3OD-01A-11R-A21T-07 TCGA-AC-A3OD         tumor         LumA</span></span>
<span><span class="co">## TCGA-AN-A0XS-01A-22R-A109-07 TCGA-AN-A0XS         tumor         LumA</span></span>
<span><span class="co">## TCGA-A2-A0EM-01A-11R-A034-07 TCGA-A2-A0EM         tumor         LumA</span></span>
<span><span class="co">## TCGA-AR-A24O-01A-11R-A169-07 TCGA-AR-A24O         tumor         LumA</span></span>
<span><span class="co">## TCGA-D8-A4Z1-01A-21R-A266-07 TCGA-D8-A4Z1         tumor         LumA</span></span></code></pre>
<div class="note">
<p>Note that there are many tutorials online that outline how to
generate expression matrices for use with differential expression and
analysis, such as the one that is returned from the
<code>exampleExpressionSet</code> function. Summarizing assay data into
such a format is out of scope for this vignette, but you can reference
the <a href="http://bioconductor.org/packages/release/data/experiment/vignettes/airway/inst/doc/airway.html" class="external-link">airway
vignette</a> for full details (among others).</p>
</div>
</div>
<div class="section level3">
<h3 id="data-analysis">Data Analysis<a class="anchor" aria-label="anchor" href="#data-analysis"></a>
</h3>
<p>We will identify the genes and genesets that are differentially
expressed between the basal and her2 subtypes. The <code>vm</code>
object has already been <code>voom</code>d using this design:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vm</span><span class="op">$</span><span class="va">design</span></span></code></pre></div>
<pre><code><span><span class="co">##                              Basal Her2 LumA</span></span>
<span><span class="co">## TCGA-A2-A0CM-01A-31R-A034-07     1    0    0</span></span>
<span><span class="co">## TCGA-BH-A0RX-01A-21R-A084-07     1    0    0</span></span>
<span><span class="co">## TCGA-BH-A18Q-01A-12R-A12D-07     1    0    0</span></span>
<span><span class="co">## TCGA-B6-A0RU-01A-11R-A084-07     1    0    0</span></span>
<span><span class="co">## TCGA-BH-A18P-01A-11R-A12D-07     0    1    0</span></span>
<span><span class="co">## TCGA-C8-A275-01A-21R-A16F-07     0    1    0</span></span>
<span><span class="co">## TCGA-C8-A12Z-01A-11R-A115-07     0    1    0</span></span>
<span><span class="co">## TCGA-A2-A0T1-01A-21R-A084-07     0    1    0</span></span>
<span><span class="co">## TCGA-AC-A3OD-01A-11R-A21T-07     0    0    1</span></span>
<span><span class="co">## TCGA-AN-A0XS-01A-22R-A109-07     0    0    1</span></span>
<span><span class="co">## TCGA-A2-A0EM-01A-11R-A034-07     0    0    1</span></span>
<span><span class="co">## TCGA-AR-A24O-01A-11R-A169-07     0    0    1</span></span>
<span><span class="co">## TCGA-D8-A4Z1-01A-21R-A266-07     0    0    1</span></span>
<span><span class="co">## attr(,"assign")</span></span>
<span><span class="co">## [1] 1 1 1</span></span>
<span><span class="co">## attr(,"contrasts")</span></span>
<span><span class="co">## attr(,"contrasts")$PAM50subtype</span></span>
<span><span class="co">## [1] "contr.treatment"</span></span></code></pre>
<p>We can test for differences between basla and her2 subtypes using the
following contrast:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">cm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html" class="external-link">makeContrasts</a></span><span class="op">(</span>BvH<span class="op">=</span><span class="va">Basal</span> <span class="op">-</span> <span class="va">Her2</span>, levels<span class="op">=</span><span class="va">vm</span><span class="op">$</span><span class="va">design</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        Contrasts</span></span>
<span><span class="co">## Levels  BvH</span></span>
<span><span class="co">##   Basal   1</span></span>
<span><span class="co">##   Her2   -1</span></span>
<span><span class="co">##   LumA    0</span></span></code></pre>
<div class="section level4">
<h4 id="differential-gene-expression">Differential Gene Expression<a class="anchor" aria-label="anchor" href="#differential-gene-expression"></a>
</h4>
<p>In this section, we first show you the straightforward analysis you
would do if you were only testing for differential gene expression.</p>
<p>With the data we have at hand, you would simply do the following:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">vm</span>, <span class="va">vm</span><span class="op">$</span><span class="va">design</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html" class="external-link">contrasts.fit</a></span><span class="op">(</span><span class="va">cm</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="va">eBayes</span></span>
<span><span class="va">tt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></span><span class="op">(</span><span class="va">fit</span>, <span class="st">'BvH'</span>, n<span class="op">=</span><span class="cn">Inf</span>, sort.by<span class="op">=</span><span class="st">'none'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="gene-set-enrichment-analysis">Gene Set Enrichment Analysis<a class="anchor" aria-label="anchor" href="#gene-set-enrichment-analysis"></a>
</h4>
<p>Given that we now have all of the pieces of data required for a
differential expression analysis, performing GSEA is trivial using the
<code>seas</code> wrapper function. We simply need to now define (1) the
battery of gene sets we want to test against, and (2) the GSEA methods
we want to explore.</p>
</div>
<div class="section level4">
<h4 id="gene-sets-to-test">Gene Sets to Test<a class="anchor" aria-label="anchor" href="#gene-sets-to-test"></a>
</h4>
<p>The <em>sparrow</em> package provides a <code>GeneSetDb</code> class
to store collections of gene sets. The <code>GeneSetDb</code> object is
used heavily for the internal functionality of <a href="https://github.com/lianos/sparrow" class="external-link">sparrow</a>,
however you can provide sparrow with collections of gene sets using
other containers from the bioconductor universe, like a
<code><a href="https://rdrr.io/pkg/BiocSet/man/BiocSet-class.html" class="external-link">BiocSet::BiocSet</a></code> or a
<code><a href="https://rdrr.io/pkg/GSEABase/man/GeneSetCollection-methods.html" class="external-link">GSEABase::GeneSetCollection</a></code>. This package provides
convenience methods to convert between these different types of gene set
containers. Please refer to <em><a href="#the-genesetdb-class">The
GeneSetDb Class</a></em> section for more details.</p>
<p>The {sparrow} package also provides convenience methods to retrieve
gene set collections from different sourckes, like <a href="http://software.broadinstitute.org/gsea/msigdb/" class="external-link">MSigDB</a>, <a href="http://pantherdb.org" class="external-link">PANTHER</a>, <a href="https://www.genome.jp/kegg/" class="external-link">KEGG</a>, etc. These methods are
named using the following pattern:
<code>get&lt;CollectionName&gt;Collection()</code> to return a
<code>BiocSet</code> with the gene sets from the collection, or
<code>get&lt;CollectionName&gt;GeneSetDb()</code> to get a
<code>GeneSetDb</code> of the same.</p>
<p>We’ll use the <code>getMSigGeneSetDb</code> convenience function
provided by the <em>sparrow</em> package to load the <a href="http://www.cell.com/cell-systems/abstract/S2405-4712(15)00218-5" class="external-link">hallmark</a>
(<code>"h"</code>) and <a href="http://software.broadinstitute.org/gsea/msigdb/collections.jsp#C2" class="external-link">c2
(curated)</a> (<code>"c2"</code>) gene set collections from <a href="http://software.broadinstitute.org/gsea/msigdb/" class="external-link">MSigDB</a>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMSigCollection.html">getMSigGeneSetDb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"h"</span>, <span class="st">"c2"</span><span class="op">)</span>, <span class="st">"human"</span>, id.type <span class="op">=</span> <span class="st">"entrez"</span><span class="op">)</span></span></code></pre></div>
<div class="note">
<p>To retrieve a <code>BiocSet</code> of these same collections, you
could do:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bsc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMSigCollection.html">getMSigCollection</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"h"</span>, <span class="st">"c2"</span><span class="op">)</span>, <span class="st">"human"</span>, id.type <span class="op">=</span> <span class="st">"entrez"</span><span class="op">)</span></span></code></pre></div>
</div>
<p>You can view a table of the gene sets defined inside a
<code>GeneSetDb</code> (<code>gdb</code>)object via its
<code>geneSets(gdb)</code> accessor:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/geneSets.html">geneSets</a></span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="va">head</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   collection                             name active  N</span></span>
<span><span class="co">## 1         C2         ABBUD_LIF_SIGNALING_1_DN  FALSE 28</span></span>
<span><span class="co">## 2         C2         ABBUD_LIF_SIGNALING_1_UP  FALSE 43</span></span>
<span><span class="co">## 3         C2         ABBUD_LIF_SIGNALING_2_DN  FALSE  7</span></span>
<span><span class="co">## 4         C2         ABBUD_LIF_SIGNALING_2_UP  FALSE 13</span></span>
<span><span class="co">## 5         C2       ABDELMOHSEN_ELAVL4_TARGETS  FALSE 16</span></span>
<span><span class="co">## 6         C2 ABDULRAHMAN_KIDNEY_CANCER_VHL_DN  FALSE 13</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="running-sparrow">Running sparrow<a class="anchor" aria-label="anchor" href="#running-sparrow"></a>
</h3>
<p>Performing multiple gene set enrichment analyses over your contrast
of interest simply requires you to provide a <code>GeneSetDb</code> (or
<code>BiocSet</code>) object along with your data and an enumeration of
the methods you want to use in your analysis.</p>
<p>The call to <code><a href="../reference/seas.html">seas()</a></code> will perform these analyses and
return a <code>SparrowResult</code> object which you can then use for
downstream analysis.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seas.html">seas</a></span><span class="op">(</span></span>
<span>  <span class="va">vm</span>, <span class="va">gdb</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'camera'</span>, <span class="st">'fry'</span>, <span class="st">'ora'</span><span class="op">)</span>,</span>
<span>  design <span class="op">=</span> <span class="va">vm</span><span class="op">$</span><span class="va">design</span>, contrast <span class="op">=</span> <span class="va">cm</span><span class="op">[</span>, <span class="st">'BvH'</span><span class="op">]</span>,</span>
<span>  <span class="co"># these parameters define which genes are differentially expressed</span></span>
<span>  feature.max.padj <span class="op">=</span> <span class="fl">0.05</span>, feature.min.logFC <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  <span class="co"># for camera:</span></span>
<span>  inter.gene.cor <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  <span class="co"># specifies the numeric covariate to bias-correct for</span></span>
<span>  <span class="co"># "size" is found in the vm$genes data.frame, which makes its way to the</span></span>
<span>  <span class="co"># internal DGE statistics table ... more on that later</span></span>
<span>  feature.bias <span class="op">=</span> <span class="st">"size"</span><span class="op">)</span></span></code></pre></div>
<p>We will unpack the details of the <code><a href="../reference/seas.html">seas()</a></code> call shortly
…</p>
</div>
<div class="section level3">
<h3 id="implicit-differential-expression">Implicit Differential Expression<a class="anchor" aria-label="anchor" href="#implicit-differential-expression"></a>
</h3>
<p>First, let’s note that in addition to running a plethora of GSEA’s
over our data we’ve also run a standard differential expression
analysis. If you’ve passed a <code>matrix</code>,
<code>ExpressionSet</code> or <code>EList</code> into
<code><a href="../reference/seas.html">seas()</a></code>, a <em>limma</em>-based
<code>lmFit  %&gt;% (eBayes|treat) %&gt;% (topTable|topTreat)</code>
pipeline was run. If a <code>DGEList</code> was passed, then
<code>seas</code> utilizes the <em>edgeR</em>-based
<code>glmQLFit %&gt;% (glmQLFTest | glmTreat) %&gt;% topTags</code>
pipeline.</p>
<p>The result of the internally run differential expression analysis is
accessible via a call to <code>logFC</code> function on the
<code>SparrowResult</code> object:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lfc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/logFC.html">logFC</a></span><span class="op">(</span><span class="va">mg</span><span class="op">)</span></span>
<span><span class="va">lfc</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">symbol</span>, <span class="va">entrez_id</span>, <span class="va">logFC</span>, <span class="va">t</span>, <span class="va">pval</span>, <span class="va">padj</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="va">head</span></span></code></pre></div>
<pre><code><span><span class="co">##         symbol entrez_id       logFC           t      pval      padj</span></span>
<span><span class="co">## 1         A1BG         1  0.67012895  1.07951394 0.2982819 0.6858344</span></span>
<span><span class="co">## 2          ADA       100  0.53844094  0.92401125 0.3708544 0.7415607</span></span>
<span><span class="co">## 3         CDH2      1000 -0.08180996 -0.09901074 0.9225083 0.9795974</span></span>
<span><span class="co">## 4         AKT3     10000  0.58338138  1.29502525 0.2158892 0.6125318</span></span>
<span><span class="co">## 5 LOC100009676 100009676 -0.09581391 -0.26985709 0.7911366 0.9398579</span></span>
<span><span class="co">## 6         MED6     10001  0.04505155  0.15082239 0.8822288 0.9701384</span></span></code></pre>
<p>We can confirm that the statistics generated internally in
<code><a href="../reference/seas.html">seas()</a></code> mimic our explicit analysis above by verifying that
the t-statistics generated by both approaches are identical.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comp</span> <span class="op">&lt;-</span> <span class="va">tt</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">entrez_id</span>, <span class="va">logFC</span>, <span class="va">t</span>, pval<span class="op">=</span><span class="va">P.Value</span>, padj<span class="op">=</span><span class="va">adj.P.Val</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span><span class="va">lfc</span>, by<span class="op">=</span><span class="st">'entrez_id'</span>, suffix<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'.tt'</span>, <span class="st">'.mg'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">comp</span><span class="op">$</span><span class="va">t.tt</span>, <span class="va">comp</span><span class="op">$</span><span class="va">t.mg</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>The internally performed differential expression analysis within the
<code><a href="../reference/seas.html">seas()</a></code> call can be customized almost as extensively as an
explicitly performed analysis that you would run using limma or edgeR by
sending more parameters through <code><a href="../reference/seas.html">seas()</a></code>’s <code>...</code>
argument.</p>
<p>See the <em><a href="#custom-differential-expression">Custom
Differential Expression</a></em> section further in the vignette as well
as the help available in <code><a href="../reference/calculateIndividualLogFC.html">?calculateIndividualLogFC</a></code> (which
is called inside the <code><a href="../reference/seas.html">seas()</a></code> function) for more
information.</p>
</div>
<div class="section level3">
<h3 id="explicit-gsea">Explicit GSEA<a class="anchor" aria-label="anchor" href="#explicit-gsea"></a>
</h3>
<p>We also have the results of all the GSEA analyses that we specified
to our <code>seas</code> call via the <code>methods</code>
parameter.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mg</span></span></code></pre></div>
<pre><code><span><span class="co">## SparrowResult (max FDR by collection set to 0.20%)</span></span>
<span><span class="co">## --------------------------------------------------- </span></span>
<span><span class="co">##    collection   method geneset_count sig_count sig_up sig_down</span></span>
<span><span class="co">## 1          C2   camera          6150       349    206      143</span></span>
<span><span class="co">## 2           H   camera            50         6      5        1</span></span>
<span><span class="co">## 3          C2      fry          6150        95     33       62</span></span>
<span><span class="co">## 4           H      fry            50         0      0        0</span></span>
<span><span class="co">## 5          C2      ora          6150        96     33       63</span></span>
<span><span class="co">## 6           H      ora            50         3      1        2</span></span>
<span><span class="co">## 7          C2 ora.down          6150        73      6       67</span></span>
<span><span class="co">## 8           H ora.down            50         2      0        2</span></span>
<span><span class="co">## 9          C2   ora.up          6150        24     21        3</span></span>
<span><span class="co">## 10          H   ora.up            50         0      0        0</span></span></code></pre>
<p>The table above enumerates the different GSEA methods run over each
geneset collection in the rows. The columns enumerate the number of
genesets that the collection has in total (<code>geneset_count</code>),
and how many were found significant at a given FDR, which is set to 20%
by default. The <code>show</code> command for the
<code>SparrowResult</code> object simply calls the
<code><a href="../reference/results.html">tabulateResults()</a></code> function, which you can call directly
with the value of <code>max.p</code> that you might find more
appropriate.</p>
</div>
<div class="section level3">
<h3 id="exploring-results">Exploring Results<a class="anchor" aria-label="anchor" href="#exploring-results"></a>
</h3>
<p>GSEA results can be examined interactively via the command line, or
via a shiny application. You can use the <code>resultNames</code>
function to find out what GSEA methods were run, and therefore available
to you, within the the <code>SparrowResult</code> object:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/results.html">resultNames</a></span><span class="op">(</span><span class="va">mg</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "camera"   "fry"      "ora"      "ora.down" "ora.up"</span></span></code></pre>
<p>Note that when running an “over representation analysis”
<code>"ora"</code> (or <code>"goseq"</code>), it will be run three
different ways. The tests will be run first by testing all
differentially expressed genes that meet a given set of min logFC and
max FDR thresholds, then separately for only genes that go up in your
contrast, and a third time for only the genes that go down.</p>
<p>The individual gene set statistics generated by each method are
available via the <code>result</code> function (or several can be
returned with <code>results</code>):</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cam.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/results.html">result</a></span><span class="op">(</span><span class="va">mg</span>, <span class="st">'camera'</span><span class="op">)</span></span>
<span><span class="va">cam.go.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/results.html">results</a></span><span class="op">(</span><span class="va">mg</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'camera'</span>, <span class="st">'ora.up'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>You can identify genesets with the strongest enrichment by filtering
and sorting against the appropriate columns. We can, for instance,
identify which hallmark gene sets show the strongest enrichment as
follows:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cam.res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.1</span>, <span class="va">collection</span> <span class="op">==</span> <span class="st">'H'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">mean.logFC</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">name</span>, <span class="va">n</span>, <span class="va">mean.logFC</span>, <span class="va">padj</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="va">head</span></span></code></pre></div>
<pre><code><span><span class="co">##                                 name   n mean.logFC         padj</span></span>
<span><span class="co">## 1            HALLMARK_MYC_TARGETS_V2  58  0.4461105 0.0002612790</span></span>
<span><span class="co">## 2 HALLMARK_INTERFERON_ALPHA_RESPONSE  96  0.3916716 0.0874709010</span></span>
<span><span class="co">## 3               HALLMARK_E2F_TARGETS 200  0.3465703 0.0001892151</span></span>
<span><span class="co">## 4            HALLMARK_MYC_TARGETS_V1 200  0.2092836 0.0234431144</span></span></code></pre>
<p>You can also list the members of a geneset and their individual
differential expression statistics for the contrast under test using the
<code>geneSet</code> function.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/geneSet.html">geneSet</a></span><span class="op">(</span><span class="va">mg</span>, name <span class="op">=</span> <span class="st">'HALLMARK_WNT_BETA_CATENIN_SIGNALING'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">symbol</span>, <span class="va">entrez_id</span>, <span class="va">logFC</span>, <span class="va">pval</span>, <span class="va">padj</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   symbol entrez_id      logFC       pval      padj</span></span>
<span><span class="co">## 1  HDAC5     10014  0.8984691 0.02253974 0.2522754</span></span>
<span><span class="co">## 2 CSNK1E      1454 -0.1793725 0.52104817 0.8317753</span></span>
<span><span class="co">## 3 CTNNB1      1499  0.2577554 0.54741640 0.8467181</span></span>
<span><span class="co">## 4   JAG1       182  0.7293432 0.02496690 0.2625306</span></span>
<span><span class="co">## 5   DVL2      1856  0.4921509 0.24186744 0.6362028</span></span>
<span><span class="co">## 6   DKK1     22943  0.6567652 0.66735589 0.8982828</span></span></code></pre>
<div class="note">
<p>The results provided in the table generated from a call to
<code>geneSet</code> are independant of GSEA method. The statistics
appended to the gene set members are simply the ones generated from a
differential expression analysis.</p>
</div>
</div>
<div class="section level3">
<h3 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a>
</h3>
<p><a href="https://github.com/lianos/sparrow" class="external-link">sparrow</a> provides a number of interactive plotting
facilities to explore the enrichment of a single geneset under the given
contrast. In the boxplots and density plots shown below, the log fold
changes (logFCs) (or t-statistics) for all genes under the contrast are
visualized in the “background” set, and these same values are shown for
the desired geneset under the “geneset” group.</p>
<p>The logFC (or t-statistics) of the genes in the gene set are plotted
as points, which allow you to hover to identify the identity of the
genes that land in the regions of the distributions you care about.</p>
<div class="note">
<p>Including interactive plots increases the size of the vignette’s by
<em>a lot</em> and will be rejected by the bioconductor build servers,
so all plots included in this vignette are static snapshots of the
javascript enabled plots you would normally get from
<code><a href="../reference/iplot.html">iplot()</a></code>.</p>
</div>
<p><strong>Boxplot</strong></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/iplot.html">iplot</a></span><span class="op">(</span><span class="va">mg</span>, <span class="st">'HALLMARK_WNT_BETA_CATENIN_SIGNALING'</span>,</span>
<span>      type <span class="op">=</span> <span class="st">'boxplot'</span>, value <span class="op">=</span> <span class="st">'logFC'</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="images/static-iplot-wnt-beta-box.png" alt="boxplot of geneset log2FC’s"><div class="figcaption">boxplot of geneset log2FC’s</div>
</div>
<p><strong>Density</strong></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/iplot.html">iplot</a></span><span class="op">(</span><span class="va">mg</span>, <span class="st">'HALLMARK_WNT_BETA_CATENIN_SIGNALING'</span>,</span>
<span>      type <span class="op">=</span> <span class="st">'density'</span>, value <span class="op">=</span> <span class="st">'logFC'</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="images/static-iplot-wnt-beta-density.png" alt="density plot of geneset log2FC’s"><div class="figcaption">density plot of geneset log2FC’s</div>
</div>
<p><strong>GSEA plot</strong></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/iplot.html">iplot</a></span><span class="op">(</span><span class="va">mg</span>, <span class="st">'HALLMARK_WNT_BETA_CATENIN_SIGNALING'</span>,</span>
<span>      type <span class="op">=</span> <span class="st">'gsea'</span>, value <span class="op">=</span> <span class="st">'logFC'</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="images/static-iplot-wnt-beta-gsea.png" alt="gsea plot of geneset log2FC’s"><div class="figcaption">gsea plot of geneset log2FC’s</div>
</div>
<div class="section level4">
<h4 id="interactive-exploration">Interactive Exploration<a class="anchor" aria-label="anchor" href="#interactive-exploration"></a>
</h4>
<p>A sister <a href="https://github.com/lianos/sparrow.shiny" class="external-link"><code>{sparrow.shiny}</code></a>
package is available that can be used to interactively explore
<code>SparrowResult</code> objects to help you try to make sense of the
enrichment hits you get (or not!). The application can be invoked as
follows:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sparrow.shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/sparrow.shiny/man/explore.html" class="external-link">explore</a></span><span class="op">(</span><span class="va">mg</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="images/sparrow-shiny.png" alt="Screen shot of interactive sparrow exploration"><div class="figcaption">Screen shot of interactive sparrow
exploration</div>
</div>
<p>Please refer to the <code>"sparrow-shiny"</code> vignette in the <a href="https://github.com/lianos/sparrow.shiny" class="external-link"><code>{sparrow.shiny}</code></a>
package for documentation on the application’s use.</p>
<div class="note">
<p>The <code>{sparrow.shiny}</code> package is currently only available
to install from GitHub, but will be available through Bioconductor
soon.</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="singe-sample-gene-set-scoring">Singe Sample Gene Set Scoring<a class="anchor" aria-label="anchor" href="#singe-sample-gene-set-scoring"></a>
</h2>
<p>It can be both convenient and effective to transform a gene-by-sample
expression matrix to a geneset-by-sample expression matrix. By doing so,
so we can quickly identify biological processes that are up/down
regulated (loosely speaking) in each sample.</p>
<p>We can generate single sample gene set scores using the gene sets
defined in a <code>GeneSetDb</code> using the
<code>scoreSingleSamples</code> function. This function takes a
<code>GeneSetDb</code>, an expression container, and a
<code>methods</code> argument, which is analagous to the
<code>methods</code> argument in the <code><a href="../reference/seas.html">seas()</a></code> call: it
defines all of the scoring methos the user wants to apply to each
sample.</p>
<p>Let’s pick a few gene sets to score our samples with for this
exercise. We’ll take the significant hallmark gene sets, or any other
significant gene set that has a large (on average) log fold change
between conditions.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sig.res</span> <span class="op">&lt;-</span> <span class="va">cam.res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"HALLMARK"</span>, <span class="va">name</span><span class="op">)</span> <span class="op">|</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">mean.logFC</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gdb.sub</span> <span class="op">&lt;-</span> <span class="va">gdb</span><span class="op">[</span><span class="fu"><a href="../reference/geneSets.html">geneSets</a></span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span><span class="op">$</span><span class="va">name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">sig.res</span><span class="op">$</span><span class="va">name</span><span class="op">]</span></span></code></pre></div>
<div class="note">
<p>Refer to the <a href="#subsetting-a-genesetdb">Subsetting a
GeneSetDb</a> section to learn how to subset a <code>GeneSetDb</code>
object to create a derivative object with fewer gene sets.</p>
</div>
<p>Recall that the GSEA analysis we performed was perfomed between the
Basal and Her2 subtypes, so we will use an expression matrix that only
has the samples from those two groups.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vm.bh</span> <span class="op">&lt;-</span> <span class="va">vm</span><span class="op">[</span>, <span class="va">vm</span><span class="op">$</span><span class="va">targets</span><span class="op">$</span><span class="va">PAM50subtype</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Basal"</span>, <span class="st">"Her2"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<div class="section level3">
<h3 id="generating-single-sample-gene-set-scores">Generating Single Sample Gene Set Scores<a class="anchor" aria-label="anchor" href="#generating-single-sample-gene-set-scores"></a>
</h3>
<p>Once we have a <code>GeneSetDb</code> object that contains all of the
gene sets we wish to use to create single sample gene set scores, we can
use the <code>scoreSingleSamples</code> function to produce these scores
using a variety of algorithmes, which the user species using the
<code>methods</code> parameter.</p>
<p>The <code>scoreSingleSamples</code> function will return a long
<code>data.frame</code> with <code>length(methods) * ncol(exprs)</code>
rows. Each row represents the score for the given <code>sample</code>
using the specified <code>method</code>. You can subset against the
<code>method</code> column to extract all of the single sample scores
for a given method.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scoreSingleSamples.html">scoreSingleSamples</a></span><span class="op">(</span><span class="va">gdb.sub</span>, <span class="va">vm.bh</span>,</span>
<span>                             methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'ewm'</span>, <span class="st">'ssgsea'</span>, <span class="st">'zscore'</span><span class="op">)</span>,</span>
<span>                             ssgsea.norm <span class="op">=</span> <span class="cn">TRUE</span>, unscale<span class="op">=</span><span class="cn">FALSE</span>, uncenter<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>                             as.dt <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Calculating ranks..."</span></span>
<span><span class="co">## [1] "Calculating absolute values from ranks..."</span></span>
<span><span class="co">## [1] "Normalizing..."</span></span></code></pre>
<p>We can see how the scores from different methods compare to each
other:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We miss you, reshape2::acast</span></span>
<span><span class="va">sw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/dcast.data.table.html" class="external-link">dcast</a></span><span class="op">(</span><span class="va">scores</span>, <span class="va">name</span> <span class="op">+</span> <span class="va">sample_id</span> <span class="op">~</span> <span class="va">method</span>, value.var<span class="op">=</span><span class="st">"score"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/corplot.html">corplot</a></span><span class="op">(</span><span class="va">sw</span><span class="op">[</span>, <span class="op">-</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>, with <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>, cluster<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="sparrow_files/figure-html/sss-pairs-1.png" width="700"></p>
<p>It is, perhaps, interesting to compare how the <code>ewm</code>
method scores change when we choose not to “uncenter” and “unscale”
them:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ewmu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scoreSingleSamples.html">scoreSingleSamples</a></span><span class="op">(</span><span class="va">gdb.sub</span>, <span class="va">vm.bh</span>,methods <span class="op">=</span> <span class="st">"ewm"</span>,</span>
<span>                           unscale <span class="op">=</span> <span class="cn">TRUE</span>, uncenter <span class="op">=</span> <span class="cn">TRUE</span>, as.dt <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ewmu</span><span class="op">[</span>, <span class="va">method</span> <span class="op">:=</span> <span class="st">"ewm_unscale"</span><span class="op">]</span></span>
<span><span class="va">scores.all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">scores</span>, <span class="va">ewmu</span><span class="op">)</span></span>
<span><span class="va">swa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/dcast.data.table.html" class="external-link">dcast</a></span><span class="op">(</span><span class="va">scores.all</span>, <span class="va">name</span> <span class="op">+</span> <span class="va">sample_id</span> <span class="op">~</span> <span class="va">method</span>, value.var<span class="op">=</span><span class="st">"score"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/corplot.html">corplot</a></span><span class="op">(</span><span class="va">swa</span><span class="op">[</span>, <span class="op">-</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>, with <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>, cluster<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="sparrow_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>Further exposition on the “ewm” (eigenWeightedMean) scoring method
can be found in the <code><a href="../reference/eigenWeightedMean.html">?eigenWeightedMean</a></code> function.</p>
</div>
<div class="section level3">
<h3 id="visualizing-single-sample-gene-set-scores">Visualizing Single Sample Gene Set Scores<a class="anchor" aria-label="anchor" href="#visualizing-single-sample-gene-set-scores"></a>
</h3>
<p>The “long” data.frame nature of the results produced by
<code>scoreSingleSamples</code> makes it convenient to use with graphing
libraries like ggplot2 so that we can create arbitrary visualizations.
Creating boxplots for gene sets per subtype is an easy way to explore
these results.</p>
<p>Let’s annotate each row in <code>scores.all</code> with the subtype
annotation and observe how these methods score each sample for a few
gene sets.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all.scores</span> <span class="op">&lt;-</span> <span class="va">scores.all</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">vm.bh</span><span class="op">$</span><span class="va">targets</span>, sample_id<span class="op">=</span><span class="va">Sample_ID</span>, subtype<span class="op">=</span><span class="va">PAM50subtype</span><span class="op">)</span>,</span>
<span>             by <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">some.scores</span> <span class="op">&lt;-</span> <span class="va">all.scores</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">all.scores</span><span class="op">$</span><span class="va">name</span><span class="op">)</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">some.scores</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">subtype</span>, <span class="va">score</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>outlier.shape<span class="op">=</span><span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>width<span class="op">=</span><span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="va">name</span> <span class="op">~</span> <span class="va">method</span><span class="op">)</span></span></code></pre></div>
<p><img src="sparrow_files/figure-html/anno-scores-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="gene-set-based-heatmap-with-mgheatmap">Gene Set Based Heatmap with mgheatmap<a class="anchor" aria-label="anchor" href="#gene-set-based-heatmap-with-mgheatmap"></a>
</h2>
<p>We often want to create expression based heatmaps that highlight the
behavior of gene sets across our samples. The <code>mgheatmap</code>
function uses the <a href="http://bioconductor.org/packages/ComplexHeatmap" class="external-link">ComplexHeatmap</a>
package to create two different types of heatmaps:</p>
<ol style="list-style-type: decimal">
<li>Gene based heatmaps, that split the genes (rows) based on their
genesets</li>
<li>Single sample gene set based heatmaps, optionally split by gene set
collection.</li>
</ol>
<p>The <code>mgheatmap</code> function has a set of arguments that
customize how the heatmap is to be created (gene level vs. gene set
level, whether to split it, etcv) and will also use the <code>...</code>
argument to pass any parameters down to the inner
<code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap</a></code> function call and customize its
behavior. The <code>mgheatmap</code> function returns a
<code>ComplexHeatmap,Heatmap</code> object for plotting or combining
with other ComplexHeatmap heatmaps or annotations in order to create
arbitrarily complex/informative heatmap figures.</p>
<div class="section level3">
<h3 id="gene-level-based-heatmap-from-genesets">Gene level based heatmap (from genesets)<a class="anchor" aria-label="anchor" href="#gene-level-based-heatmap-from-genesets"></a>
</h3>
<p>You can plot a heatmap of the genes from a predefined set of gene
sets by providing the gene sets you want to visualize in a
<code>GeneSetDb</code> object.</p>
<p>We’ll create a new <code>GeneSetDb</code> object using the first two
gene sets in <code>gdb.sub</code> and draw a heatmap of their
expression.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gs.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneSets.html">geneSets</a></span><span class="op">(</span><span class="va">gdb.sub</span><span class="op">)</span></span>
<span><span class="va">gdb.2</span> <span class="op">&lt;-</span> <span class="va">gdb.sub</span><span class="op">[</span><span class="fu"><a href="../reference/geneSets.html">geneSets</a></span><span class="op">(</span><span class="va">gdb.sub</span><span class="op">)</span><span class="op">$</span><span class="va">name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gs.sub</span><span class="op">$</span><span class="va">name</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">col.anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html" class="external-link">HeatmapAnnotation</a></span><span class="op">(</span></span>
<span>  df <span class="op">=</span> <span class="va">vm.bh</span><span class="op">$</span><span class="va">targets</span><span class="op">[</span>, <span class="st">'PAM50subtype'</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>PAM50subtype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Basal <span class="op">=</span> <span class="st">"gray"</span>, Her2 <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/mgheatmap.html">mgheatmap</a></span><span class="op">(</span><span class="va">vm.bh</span>, <span class="va">gdb.2</span>, aggregate.by <span class="op">=</span> <span class="st">"none"</span>, split <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>          show_row_names <span class="op">=</span> <span class="cn">FALSE</span>, show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          recenter <span class="op">=</span> <span class="cn">TRUE</span>, top_annotation <span class="op">=</span> <span class="va">col.anno</span>, zlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="sparrow_files/figure-html/gheatmap-1.png" width="384"></p>
</div>
<div class="section level3">
<h3 id="gene-set-based-heatmap">Gene set-based heatmap<a class="anchor" aria-label="anchor" href="#gene-set-based-heatmap"></a>
</h3>
<p>You can often get a higher information:ink ratio by plotting heatmaps
based on single sample gene set scores as opposed to the genes that make
up a geneset.</p>
<p>Let’s see what the simple 2-geneset version of the heatmap above
looks like:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mgheatmap.html">mgheatmap</a></span><span class="op">(</span><span class="va">vm.bh</span>, <span class="va">gdb.2</span>, aggregate.by <span class="op">=</span> <span class="st">"ewm"</span>, split <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          show_row_names <span class="op">=</span> <span class="cn">TRUE</span>, show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          top_annotation <span class="op">=</span> <span class="va">col.anno</span><span class="op">)</span></span></code></pre></div>
<p><img src="sparrow_files/figure-html/gshm2-1.png" width="768"></p>
<p>Plotted in this way, we can now show the activity of a greater number
of genesets</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mgheatmap.html">mgheatmap</a></span><span class="op">(</span><span class="va">vm.bh</span>, <span class="va">gdb.sub</span>,</span>
<span>          aggregate.by <span class="op">=</span> <span class="st">'ewm'</span>, split<span class="op">=</span><span class="cn">TRUE</span>, recenter <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>          show_row_names<span class="op">=</span><span class="cn">TRUE</span>, show_column_names<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>          top_annotation<span class="op">=</span><span class="va">col.anno</span>, zlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2.5</span>, <span class="fl">2.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="sparrow_files/figure-html/gshm-all-1.png" width="768"></p>
</div>
</div>
<div class="section level2">
<h2 id="the-genesetdb-class">The GeneSetDb Class<a class="anchor" aria-label="anchor" href="#the-genesetdb-class"></a>
</h2>
<p>The GeneSetDb class was developed to address the internal needs of
the sparrow package for fast look up, subsetting, cross reference, etc.
of a collection of gene sets. At the time (~2015), it was developed
because the classes used for this purpose in the bioconductor ecosystem
(a <code><a href="https://rdrr.io/pkg/GSEABase/man/GeneSetCollection-methods.html" class="external-link">GSEABase::GeneSetCollection</a></code>, or a simple list of gene
vectors) didn’t cut the mustard.</p>
<p>More recently, bioc-core has developed a new class called a
<code>BiocSet</code> that is feature-rich and shares significant overlap
with the features in the <code><a href="../reference/GeneSetDb-class.html">sparrow::GeneSetDb</a></code> class. Although
we can’t quite replace the internals of {sparrow} to use the
<code>BiocSet</code> just yet, users are encouraged to provide
collections of gene sets in the form of a <code>BiocSet</code>
everywhere {sparrow} functions require gene set collections, like
<code><a href="../reference/seas.html">seas()</a></code> and <code><a href="../reference/scoreSingleSamples.html">scoreSingleSamples()</a></code>. You can also
convert a <code><a href="../reference/GeneSetDb-class.html">sparrow::GeneSetDb()</a></code> to a <code>BiocSet</code>
via a simple call: <code>as(gdb, "BiocSet")</code>.</p>
<p>The remainder of this section provides a quick overview of the
<code>GeneSetDb</code> class.</p>
<p>The GeneSetDb object uses the <code>data.table</code> package
internally for fast lookup. Internally the collection of gene set
information is minimally stored as a three-column
<code>data.table</code> in “long form”, which has the following
columns:</p>
<ul>
<li>collection</li>
<li>name</li>
<li>feature_id</li>
</ul>
<p>More columns can be added to the internal <code>data.table</code> (a
“symbol” column, for instance), but those are the only three you
need.</p>
<p>To see what we are talking about, exactly, you can call the
<code>as.data.frame</code> function on a <code>GeneSetDb</code>
object:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/conversion.html">as.data.frame</a></span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">201</span><span class="op">:</span><span class="fl">205</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##     collection                     name feature_id   symbol</span></span>
<span><span class="co">## 1           C2 ABBUD_LIF_SIGNALING_1_DN  100133941     CD24</span></span>
<span><span class="co">## 2           C2 ABBUD_LIF_SIGNALING_1_DN      10753    CAPN9</span></span>
<span><span class="co">## 3           C2 ABBUD_LIF_SIGNALING_1_DN     146556 C16orf89</span></span>
<span><span class="co">## 4           C2 ABBUD_LIF_SIGNALING_1_DN       1644      DDC</span></span>
<span><span class="co">## 5           C2 ABBUD_LIF_SIGNALING_1_DN       1943    EFNA2</span></span>
<span><span class="co">## 201         C2    ABE_VEGFA_TARGETS_2HR       3949     LDLR</span></span>
<span><span class="co">## 202         C2    ABE_VEGFA_TARGETS_2HR       4171     MCM2</span></span>
<span><span class="co">## 203         C2    ABE_VEGFA_TARGETS_2HR       5055 SERPINB2</span></span>
<span><span class="co">## 204         C2    ABE_VEGFA_TARGETS_2HR       5133    PDCD1</span></span>
<span><span class="co">## 205         C2    ABE_VEGFA_TARGETS_2HR       5493      PPL</span></span></code></pre>
<p>The <code>(collection,name)</code> tuple is the primary key of a gene
set. The <code>feature_id</code> column stores gene identifiers. For the
time being, it will be most natural for these IDs to simply be ensembl
gene identifiers (or entrez ids) as many of the annotation databases use
these identifiers, as well. In reality, you will want the values in the
<code>feature_id</code> columns to match with the feature id’s you have
in your data container (ie. the <code><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames()</a></code> of a
<code>SummarizedExperiment</code>, for instance).</p>
<div class="section level3">
<h3 id="building-a-genesetdb">Building a GeneSetDb<a class="anchor" aria-label="anchor" href="#building-a-genesetdb"></a>
</h3>
<p>The sparrow package provides convenience functions to fetch genesets
from many sources and convert them into a GeneSetDb object. The two most
useful sources may be:</p>
<ul>
<li>
<a href="http://software.broadinstitute.org/gsea/msigdb/" class="external-link">MSigDB</a>
via <code>getMSigGeneSetDb(...)</code>. Although the core
<code>sparrow</code> package provides the getter function for these
genesets, the main data retrieval functionality is provided through the
<a href="https://cran.r-project.org/package=msigdbr" class="external-link">msigdbr</a>
package.</li>
<li>
<a href="http://pantherdb.org" class="external-link">PANTHER</a> (pathways and GOSLIM) via
<code><a href="../reference/getPantherCollection.html">getPantherGeneSetDb()</a></code>
</li>
<li>
<a href="https://www.genome.jp/kegg/" class="external-link">KEGG</a> via
<code>getKeggGeneSetDb(...)</code>
</li>
</ul>
<p>We also provide similarly named methos to retrieve these gene set
collections as a <code>BiocSet</code>, just substitute
<code>"Collection"</code> for <code>"GeneSetDb"</code>, ie.
<code>getMsigCollection(...)</code>,
<code>getPantherCollection(...)</code>, and
<code>getKeggCollection(...)</code>.</p>
<p>You can create a custom <code>GeneSetDb</code> via the
<code><a href="../reference/GeneSetDb-class.html">GeneSetDb()</a></code> constructor, which accepts the following types
of inputs.</p>
<ol style="list-style-type: decimal">
<li>A <code>BiocSet</code>
</li>
<li>A <code>GeneSetCollection</code>
</li>
<li>A data.frame of geneset membership. This requires
<code>collection</code>, <code>name</code>, and <code>feature_id</code>
columns. Reference the output of <code>as.data.frame(gdb)</code> shown
above.</li>
<li>A named list of gene identifier vectors that represent genesets for
a single collection</li>
<li>A named list of (2)-like lists. The top level names are the names of
the different collections, and each sublist represents the genesets in
that collection.</li>
</ol>
<p>Two <code>GeneSetDb</code> objects can be combined using the
<code>cobine()</code> function. For now it is your responsibility to
ensure that the two <code>GeneSetDb</code> objects are “reasonably
conformable”, ie. they use the same types of gene identifiers, and are
referencing the same species, etc.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">msigdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMSigCollection.html">getMSigGeneSetDb</a></span><span class="op">(</span><span class="st">'H'</span>, <span class="st">'human'</span><span class="op">)</span></span>
<span><span class="va">goslimdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPantherCollection.html">getPantherGeneSetDb</a></span><span class="op">(</span><span class="st">'goslim'</span>, <span class="st">'human'</span><span class="op">)</span></span>
<span><span class="va">gdb.uber</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/combine.html" class="external-link">combine</a></span><span class="op">(</span><span class="va">msigdb</span>, <span class="va">goslimdb</span><span class="op">)</span></span></code></pre></div>
<p>See the help and examples in <code><a href="../reference/GeneSetDb-class.html">?GeneSetDb</a></code> for more
information.</p>
<div class="warning">
<p>For some reason the <code>PANTHER.db</code> package needs to be
installed in a user-writable package location for this to work properly.
If you see an error that speaks to using “rsqlite to write to a readonly
database”, you will have to re-install <code>PANTHER.db</code> in a
user-writable directory using
<code>BiocManager::install("PANTHER.db")</code></p>
</div>
</div>
<div class="section level3">
<h3 id="subsetting-a-genesetdb">Subsetting a GeneSetDb<a class="anchor" aria-label="anchor" href="#subsetting-a-genesetdb"></a>
</h3>
<p>The subsetting functionality for a <code>GeneSetDb</code> is a bit
clunky. We assume you want to subset a GeneSetDb to include a subset of,
well, gene sets.</p>
<p>One way you can do that is to provide a logical vector that is as
long as there are gene sets in the GeneSetDb as an index.</p>
<p>For instance, if we want to include only the genesets in <a href="https://www.gsea-msigdb.org/gsea/msigdb/genesets.jsp?collection=CP:PID" class="external-link">CP:PID</a>,
you can do that. This subcatory information is stored in the
<code>"subcategory"</code> column from <code>geneSets(gdb)</code></p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/geneSets.html">geneSets</a></span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span><span class="op">$</span><span class="va">subcategory</span> <span class="op">==</span> <span class="st">"CP:PID"</span></span>
<span><span class="va">gdb.sub</span> <span class="op">&lt;-</span> <span class="va">gdb</span><span class="op">[</span><span class="va">keep</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/geneSets.html">geneSets</a></span><span class="op">(</span><span class="va">gdb.sub</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   collection                           name active  N  n subcategory gs_id</span></span>
<span><span class="co">## 1         C2 PID_A6B1_A6B4_INTEGRIN_PATHWAY  FALSE 46 NA      CP:PID  M239</span></span>
<span><span class="co">## 2         C2            PID_AJDISS_2PATHWAY  FALSE 48 NA      CP:PID  M142</span></span>
<span><span class="co">## 3         C2               PID_ALK1_PATHWAY  FALSE 26 NA      CP:PID  M185</span></span>
<span><span class="co">## 4         C2               PID_ALK2_PATHWAY  FALSE 11 NA      CP:PID  M203</span></span>
<span><span class="co">## 5         C2    PID_ALPHA_SYNUCLEIN_PATHWAY  FALSE 32 NA      CP:PID  M275</span></span>
<span><span class="co">## 6         C2   PID_AMB2_NEUTROPHILS_PATHWAY  FALSE 41 NA      CP:PID  M159</span></span></code></pre>
<p>You can also subset a <code>GeneSetDb</code> to only include gene
sets that contain certain features:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gdb.sub2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subsetByFeatures.html">subsetByFeatures</a></span><span class="op">(</span><span class="va">gdb</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'10014'</span>, <span class="st">'1454'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gdb</span><span class="op">)</span>; <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">gdb.sub2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 6230</span></span></code></pre>
<pre><code><span><span class="co">## [1] 120</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="active-vs-inactive-gene-sets">Active vs Inactive Gene Sets<a class="anchor" aria-label="anchor" href="#active-vs-inactive-gene-sets"></a>
</h3>
<p>A <code>GeneSetDb</code> is used to hold “the universe” of genes that
belong to different gene sets across different collections. Depending on
the assay performed to measure these genes, the set of genes you observe
in your study will likely be a subset of the genes in the
<code>GeneSetDb</code>. As such, prior to using a <code>GeneSetDb</code>
for GSEA, it must be “conformed” to a target object that will be used
for the input to the GESA (either a matrix of expression, or a pre
ranked vector of statistics). This step will index into the target
expression object and identify which rows of the object correspond to
which genes in the <code>GeneSetDb</code>.</p>
<p>“Conformation” happens automatically within the <code><a href="../reference/seas.html">seas()</a></code>
call, but we call it explicitly below to outline its functionality. The
command below conforms the <code>GeneSetDb</code> to our target “voomed”
<code>EList</code>, and deactivates gene sets (i.e. removes them from
downstream GSEA) that have less than 10 or more than 100 genes that were
found in <code>vm</code>:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gdbc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/conform.html">conform</a></span><span class="op">(</span><span class="va">gdb</span>, <span class="va">vm</span>, min.gs.size<span class="op">=</span><span class="fl">10</span>, max.gs.size<span class="op">=</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/geneSets.html">geneSets</a></span><span class="op">(</span><span class="va">gdbc</span>, active.only<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   collection                             name active  N  n subcategory gs_id</span></span>
<span><span class="co">## 1         C2         ABBUD_LIF_SIGNALING_1_DN   TRUE 28 25         CGP M1423</span></span>
<span><span class="co">## 2         C2         ABBUD_LIF_SIGNALING_1_UP   TRUE 43 37         CGP M1458</span></span>
<span><span class="co">## 3         C2         ABBUD_LIF_SIGNALING_2_DN  FALSE  7  5         CGP M1481</span></span>
<span><span class="co">## 4         C2         ABBUD_LIF_SIGNALING_2_UP   TRUE 13 12         CGP M1439</span></span>
<span><span class="co">## 5         C2       ABDELMOHSEN_ELAVL4_TARGETS   TRUE 16 15         CGP M2509</span></span>
<span><span class="co">## 6         C2 ABDULRAHMAN_KIDNEY_CANCER_VHL_DN   TRUE 13 12         CGP M2096</span></span></code></pre>
<p>We can see that, only 23 of the 26 genes in the
<code>(C2,ABBUD_LIF_SIGNALING_1_DN)</code> were found in the rows of
<code>vm</code>, and the <code>(C2,ABBUD_LIF_SIGNALING_2_DN)</code> was
“deactivated.” Deactivated (<code>active == FALSE</code>) gene sets will
be ignored during downstream analyses. This gene set was deactivated
because it only has five “conformed” genes, but the minimum geneset size
we wanted to consider (<code>min.gs.size</code>) was set to ten in our
call to <code>conform</code>.</p>
</div>
<div class="section level3">
<h3 id="accessing-members-of-a-gene-set">Accessing members of a gene set<a class="anchor" aria-label="anchor" href="#accessing-members-of-a-gene-set"></a>
</h3>
<p>The <code>geneSet</code> and <code>featureIds</code> functions allow
the user to identify the genes found in a gene set. Both of these
functions take an <code>active.only</code> argument, which is
<code>TRUE</code> by default. This specifies that only the genes that
have been successfully conformed to a gene set should be the ones that
are returned.</p>
<p>For instance, we can identify which genes belong to the
<code>(C2,ABBUD_LIF_SIGNALING_1_DN)</code>, and which three were not
found in <code>vm</code> like so:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">missed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">setdiff</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/featureIds.html">featureIds</a></span><span class="op">(</span><span class="va">gdbc</span>, <span class="st">'C2'</span>, <span class="st">'ABBUD_LIF_SIGNALING_1_DN'</span>, active.only<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/featureIds.html">featureIds</a></span><span class="op">(</span><span class="va">gdbc</span>, <span class="st">'C2'</span>, <span class="st">'ABBUD_LIF_SIGNALING_1_DN'</span>, active.only<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">missed</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "1644" "1943" "3170"</span></span></code></pre>
<p>or we can use the <code>geneSet</code> function to return a
<code>data.frame</code> of these results:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gdbc</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/geneSet.html">geneSet</a></span><span class="op">(</span><span class="st">'C2'</span>, <span class="st">'ABBUD_LIF_SIGNALING_1_DN'</span>, active.only <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">feature_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">missed</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    collection                     name active  N  n feature_id symbol</span></span>
<span><span class="co">## 4          C2 ABBUD_LIF_SIGNALING_1_DN   TRUE 28 25       1644    DDC</span></span>
<span><span class="co">## 5          C2 ABBUD_LIF_SIGNALING_1_DN   TRUE 28 25       1943  EFNA2</span></span>
<span><span class="co">## 12         C2 ABBUD_LIF_SIGNALING_1_DN   TRUE 28 25       3170  FOXA2</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="mapping-of-gene-set-featureids-to-target-expression-containers">Mapping of gene set featureIds to target expression containers<a class="anchor" aria-label="anchor" href="#mapping-of-gene-set-featureids-to-target-expression-containers"></a>
</h3>
<p>It may be that the IDs used in a gene set collection are different
from the ones used as the rownames of your expression container. For
instance, the IDs used for a given gene set collection in the
<code>GeneSetDb</code> might be Ensembl gene identifiers, but the
rownames of the expression object might be Entrez ID. This is where the
<code>mapping</code> parameter becomes useful.</p>
<p>The <code>GeneSetDb</code> class has a concept of an internal
<code>featureIdMap</code> to accommodate these scenarios, which would
allow for a non-destructive mapping of the original IDs to a new “ID
space” (entrez to ensembl, for instance).</p>
<p>This functionality is not ready for this release, but it’s just a
note to keep the user aware of some future development of the package.
For the time being, the user is required to manually map the feautreIds
in their expression matrix to be concordant with the ones found in the
<code>GeneSetDb</code>.</p>
<p><strong>In the meantime</strong>, a <code>renameRows</code>
convenience function is provided here to easily rename the rows of our
expression container to different values. For instance, to rename this
is how you might rename the rows of your assay container to use
symbols:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/examples.html">exampleExpressionSet</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">vms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/renameRows.html">renameRows</a></span><span class="op">(</span><span class="va">vm</span>, <span class="st">"symbol"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">vm</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">vms</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      [,1]        [,2]          </span></span>
<span><span class="co">## [1,] "1"         "A1BG"        </span></span>
<span><span class="co">## [2,] "100"       "ADA"         </span></span>
<span><span class="co">## [3,] "1000"      "CDH2"        </span></span>
<span><span class="co">## [4,] "10000"     "AKT3"        </span></span>
<span><span class="co">## [5,] "100009676" "LOC100009676"</span></span>
<span><span class="co">## [6,] "10001"     "MED6"</span></span></code></pre>
<p>We grabbed the <code>symbol</code> column from <code>vm$genes</code>
and “smartly” renamed the rows of <code>vm</code> with the values there.
Refer to the <code><a href="../reference/renameRows.html">?renameRows</a></code> man page for more details. This,
of course, still requires you to manually fetch and map identifiers, but
still …</p>
</div>
</div>
<div class="section level2">
<h2 id="customizing-analyses">Customizing Analyses<a class="anchor" aria-label="anchor" href="#customizing-analyses"></a>
</h2>
<p>The internal differential expression analysis as well the gene set
enrichment analyses can be customized by passing parameters through the
<code>...</code> in the <code><a href="../reference/seas.html">seas()</a></code> function.</p>
<div class="section level3">
<h3 id="custom-differential-expression">Custom Differential Expression<a class="anchor" aria-label="anchor" href="#custom-differential-expression"></a>
</h3>
<p>The internal differential expression pipeline, exported via the
<code>calculateIndividualLogFC</code> function allows the end user to
configure an “arbitrarily complex” differential expression analysis
using either edgeR’s quasilikelihood framework (if the input is a
DGEList) or a direct limma analysis (with a pre-voomed EList, expression
matrix, or whatever).</p>
<p>User’s should refer to the <code><a href="../reference/calculateIndividualLogFC.html">?calculateIndividualLogFC</a></code>
help page to see which parameters are exposed for a differential
expression analysis and configure them accordingly. When calling
<code><a href="../reference/seas.html">seas()</a></code> use these same parameters in the call and they will
be provided to <code>calculateIndividualLogFC</code>.</p>
<p>For instance, if you wanted to use limma’s “treat” functionality to
specify a minimal log fold change threshold for statistical
significance, you would do so as follows:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seas.html">seas</a></span><span class="op">(</span><span class="va">vm</span>, <span class="va">gdb</span>, <span class="st">"goseq"</span>, design <span class="op">=</span> <span class="va">vm</span><span class="op">$</span><span class="va">design</span>, <span class="va">cm</span><span class="op">[</span>, <span class="st">'BvH'</span><span class="op">]</span>,</span>
<span>           treat.lfc<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span>,</span>
<span>           <span class="co">## feature length vector required for goseq</span></span>
<span>           feature.bias<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">vm</span><span class="op">$</span><span class="va">genes</span><span class="op">$</span><span class="va">size</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">vm</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Using the internal <code>treat</code> functionality would really only
affect enrichment tests that first threshold the genes in your
experiment as “significant” or not, like <code>goseq</code> and not
tests like <code>camera</code>.</p>
</div>
<div class="section level3">
<h3 id="custom-gsea">Custom GSEA<a class="anchor" aria-label="anchor" href="#custom-gsea"></a>
</h3>
<p>The GSEA methods that are wrapped by <code><a href="../reference/seas.html">seas()</a></code> all take the
same parameters that are defined by their implementation. Simply pass
these parameters down via the <code>...</code> in the
<code><a href="../reference/seas.html">seas()</a></code> call.</p>
<p>For instance, you can read <code><a href="https://rdrr.io/pkg/limma/man/camera.html" class="external-link">?camera</a></code> to find that the
<code>camera</code> method accepts an <code>inter.gene.cor</code>
parameter, and <code><a href="https://rdrr.io/pkg/limma/man/roast.html" class="external-link">?roast</a></code> will tell you that you can specify
the number of rotations used via the <code>nrot</code> parameter.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mgx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seas.html">seas</a></span><span class="op">(</span><span class="va">vm</span>, <span class="va">gdb</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'camera'</span>, <span class="st">'roast'</span><span class="op">)</span>, </span>
<span>            design <span class="op">=</span> <span class="va">vm</span><span class="op">$</span><span class="va">design</span>, contrast <span class="op">=</span> <span class="va">cm</span><span class="op">[</span>, <span class="st">'BvH'</span><span class="op">]</span>,</span>
<span>            inter.gene.cor <span class="op">=</span> <span class="fl">0.04</span>, nrot <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h2>
<details><summary>
Session Information
</summary><div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] data.table_1.15.4     edgeR_4.3.4           limma_3.61.1         </span></span>
<span><span class="co">##  [4] circlize_0.4.16       ComplexHeatmap_2.21.0 ggplot2_3.5.1        </span></span>
<span><span class="co">##  [7] dplyr_1.1.4           magrittr_2.0.3        sparrow_1.11.0       </span></span>
<span><span class="co">## [10] BiocStyle_2.33.0     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3          jsonlite_1.8.8             </span></span>
<span><span class="co">##   [3] shape_1.4.6.1               magick_2.8.3               </span></span>
<span><span class="co">##   [5] farver_2.1.2                rmarkdown_2.27             </span></span>
<span><span class="co">##   [7] BiocSet_1.19.0              GlobalOptions_0.1.2        </span></span>
<span><span class="co">##   [9] fs_1.6.4                    BiocIO_1.15.0              </span></span>
<span><span class="co">##  [11] zlibbioc_1.51.0             ragg_1.3.2                 </span></span>
<span><span class="co">##  [13] vctrs_0.6.5                 memoise_2.0.1              </span></span>
<span><span class="co">##  [15] DelayedMatrixStats_1.27.0   htmltools_0.5.8.1          </span></span>
<span><span class="co">##  [17] S4Arrays_1.5.1              Rhdf5lib_1.27.0            </span></span>
<span><span class="co">##  [19] SparseArray_1.5.5           rhdf5_2.49.0               </span></span>
<span><span class="co">##  [21] sass_0.4.9                  bslib_0.7.0                </span></span>
<span><span class="co">##  [23] htmlwidgets_1.6.4           desc_1.4.3                 </span></span>
<span><span class="co">##  [25] plyr_1.8.9                  plotly_4.10.4              </span></span>
<span><span class="co">##  [27] cachem_1.1.0                lifecycle_1.0.4            </span></span>
<span><span class="co">##  [29] iterators_1.0.14            pkgconfig_2.0.3            </span></span>
<span><span class="co">##  [31] rsvd_1.0.5                  Matrix_1.7-0               </span></span>
<span><span class="co">##  [33] R6_2.5.1                    fastmap_1.2.0              </span></span>
<span><span class="co">##  [35] GenomeInfoDbData_1.2.12     MatrixGenerics_1.17.0      </span></span>
<span><span class="co">##  [37] clue_0.3-65                 digest_0.6.35              </span></span>
<span><span class="co">##  [39] colorspace_2.1-0            AnnotationDbi_1.67.0       </span></span>
<span><span class="co">##  [41] S4Vectors_0.43.0            irlba_2.3.5.1              </span></span>
<span><span class="co">##  [43] GenomicRanges_1.57.0        textshaping_0.3.7          </span></span>
<span><span class="co">##  [45] RSQLite_2.3.6               beachmat_2.21.0            </span></span>
<span><span class="co">##  [47] labeling_0.4.3              fansi_1.0.6                </span></span>
<span><span class="co">##  [49] httr_1.4.7                  abind_1.4-5                </span></span>
<span><span class="co">##  [51] compiler_4.4.0              bit64_4.0.5                </span></span>
<span><span class="co">##  [53] withr_3.0.0                 doParallel_1.0.17          </span></span>
<span><span class="co">##  [55] backports_1.4.1             BiocParallel_1.39.0        </span></span>
<span><span class="co">##  [57] viridis_0.6.5               DBI_1.2.2                  </span></span>
<span><span class="co">##  [59] highr_0.10                  BiasedUrn_2.0.11           </span></span>
<span><span class="co">##  [61] HDF5Array_1.33.0            DelayedArray_0.31.1        </span></span>
<span><span class="co">##  [63] rjson_0.2.21                tools_4.4.0                </span></span>
<span><span class="co">##  [65] msigdbr_7.5.1               glue_1.7.0                 </span></span>
<span><span class="co">##  [67] rhdf5filters_1.17.0         checkmate_2.3.1            </span></span>
<span><span class="co">##  [69] cluster_2.1.6               generics_0.1.3             </span></span>
<span><span class="co">##  [71] gtable_0.3.5                tidyr_1.3.1                </span></span>
<span><span class="co">##  [73] BiocSingular_1.21.0         ScaledMatrix_1.13.0        </span></span>
<span><span class="co">##  [75] utf8_1.2.4                  XVector_0.45.0             </span></span>
<span><span class="co">##  [77] BiocGenerics_0.51.0         foreach_1.5.2              </span></span>
<span><span class="co">##  [79] pillar_1.9.0                babelgene_22.9             </span></span>
<span><span class="co">##  [81] GSVA_1.53.2                 lattice_0.22-6             </span></span>
<span><span class="co">##  [83] bit_4.0.5                   annotate_1.83.0            </span></span>
<span><span class="co">##  [85] tidyselect_1.2.1            SingleCellExperiment_1.27.0</span></span>
<span><span class="co">##  [87] locfit_1.5-9.9              Biostrings_2.73.0          </span></span>
<span><span class="co">##  [89] knitr_1.46                  gridExtra_2.3              </span></span>
<span><span class="co">##  [91] bookdown_0.39               IRanges_2.39.0             </span></span>
<span><span class="co">##  [93] SummarizedExperiment_1.35.0 stats4_4.4.0               </span></span>
<span><span class="co">##  [95] xfun_0.44                   Biobase_2.65.0             </span></span>
<span><span class="co">##  [97] statmod_1.5.0               matrixStats_1.3.0          </span></span>
<span><span class="co">##  [99] UCSC.utils_1.1.0            lazyeval_0.2.2             </span></span>
<span><span class="co">## [101] yaml_2.3.8                  evaluate_0.23              </span></span>
<span><span class="co">## [103] codetools_0.2-20            tibble_3.2.1               </span></span>
<span><span class="co">## [105] BiocManager_1.30.23         graph_1.83.0               </span></span>
<span><span class="co">## [107] cli_3.6.2                   ontologyIndex_2.12         </span></span>
<span><span class="co">## [109] xtable_1.8-4                systemfonts_1.1.0          </span></span>
<span><span class="co">## [111] munsell_0.5.1               jquerylib_0.1.4            </span></span>
<span><span class="co">## [113] Rcpp_1.0.12                 GenomeInfoDb_1.41.0        </span></span>
<span><span class="co">## [115] png_0.1-8                   XML_3.99-0.16.1            </span></span>
<span><span class="co">## [117] parallel_4.4.0              pkgdown_2.0.9              </span></span>
<span><span class="co">## [119] blob_1.2.4                  sparseMatrixStats_1.17.0   </span></span>
<span><span class="co">## [121] SpatialExperiment_1.15.0    viridisLite_0.4.2          </span></span>
<span><span class="co">## [123] GSEABase_1.67.0             scales_1.3.0               </span></span>
<span><span class="co">## [125] purrr_1.0.2                 crayon_1.5.2               </span></span>
<span><span class="co">## [127] GetoptLong_1.0.5            rlang_1.1.3                </span></span>
<span><span class="co">## [129] KEGGREST_1.45.0</span></span></code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Steve Lianoglou, Aratus Informatics, LLC, Denali Therapeutics, Genentech.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
